### PR TITLE
Refactor teaEncode/teaDecode functions to use std::string_view

### PR DESCRIPTION
**Pull request by Pirulax**
This PR replaces teaEncode/teaDecode input from `std::string` to `std::string_view`. Also adds a new macro `SDK_WITH_TEA`, for compatibility reasons (Game and multiplayer SA still use c++14, and `string_view` is c++17).

Also, now it uses `std::move` in the lambda capture list for `en/decodeString` for improved performance.

### CODE REVIEW

**Comment by Pirulax** [comment]
Fail.. I totally forgot about me pushing the commit, and I soft reseted.. nvm...

File: Shared/sdk/CScriptArgReader.h [hunk 1]
```diff
@@ -586,6 +586,33 @@ class CScriptArgReader
...
+            };
+
+            break;
+        }
+        default:
+        {
+            outValue = { nullptr, 0 };
+            SetTypeError("string");
+        }
+        }
```
Comment (StrixG) [inline Shared/sdk/CScriptArgReader.h]: Corrected formatting:
```suggestion
        switch (lua_type(m_luaVM, m_iIndex))
        {
            case LUA_TNUMBER:
            case LUA_TSTRING:
            {
                outValue = {
                    lua_tostring(m_luaVM, m_iIndex),
                    lua_strlen(m_luaVM, m_iIndex)
                };

                break;
            }
            default:
            {
                outValue = { nullptr, 0 };
                SetTypeError("string");
            }
        }
```

**Comment by qaisjp** [comment]
Is it possible to `#if` based on the the C++ version instead of a new macro?

**Comment by Pirulax** [comment]
Yes indeed, it is. Good idea, not sure why I didnt do it that way.

**Comment by Pirulax** [comment]
Update: It's not, as MSVC's cpp version macro is fucked on some older versions, and I dont want to introduce inconsistencies across compilers.

**Comment by botder** [comment]
> Update: It's not, as MSVC's cpp version macro is fucked on some older versions, and I dont want to introduce inconsistencies across compilers.

Can you update this pull request, please? We've had the commit 6d4e8adb843adbb77d9627c1c74c3d90d43c96a9 (https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170) since.

**Pull request closed without merging**